sale_loyalty_partner_applicability/models/sale_order.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2023 Tecnativa - Pilar Vargas
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+import ast
+
+from odoo import _, models
+from odoo.osv import expression
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_partner_domain(self, rule):
+        # For now, let's return the rule's partner domain.
+        if rule.rule_partners_domain != "[]":
+            return ast.literal_eval(rule.rule_partners_domain)
+        else:
+            return []  # Return an empty domain if the rule's partner domain is empty.
+
+    def _is_valid_partner(self, program):
+        """
+        Check if the partner is eligible for a loyalty program based on partner domains.
+        This method iterates through the loyalty program's rules and their partner
+        domains. It verifies if the partner meets the eligibility criteria specified
+        in the partner domain of each rule. When the partner is found eligible for a
+        rule, the program is considered valid.
+        Args:
+            program (recordset): The loyalty program for which partner eligibility is checked.
+        Returns:
+            bool: True if the partner is eligible for the program, False otherwise.
+        """
+        for rule in program.rule_ids:
+            partner_domain = self._get_partner_domain(rule)
+            if partner_domain != "[]":
+                domain = expression.AND(
+                    [partner_domain, [("id", "=", self.partner_id.id)]]
```

Comment:
Well, precisely the part we're interested in is the one that's still encapsulated :sweat_smile: